src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixesAndRefactorings
+{
+    /// <summary>
+    /// Helper methods for DocumentBasedFixAllProvider common to code fixes and refactorings.
+    /// </summary>
+    internal static class DocumentBasedFixAllProviderHelpers
+    {
+        public static async Task<Solution?> FixAllContextsAsync<TFixAllContext>(
```

Comment:
Existing code just moved to this new shared helper type.

Reply:
i'm glad this pattern worked out.